### PR TITLE
feat(gcs): Add google cloud storage connector

### PR DESCRIPTION
**Pull request by ztefanie**
## Description
Adding a new google cloud storage connector.
Docs PR is here: https://github.com/camunda/camunda-docs/pull/6104

## Related issues

related to: 
https://github.com/camunda/team-connectors/issues/1077
https://github.com/camunda/product-hub/issues/2712

## Checklist

- [x] PR has a **milestone** or the `no milestone` label.
- [x] Backport labels are added if these code changes should be backported. No backport label is added to the latest release, as this branch will be rebased onto main before the next release.

### CODE REVIEW

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/ObjectStorageConnectorFunction.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    propertyGroups = {
+      @ElementTemplate.PropertyGroup(id = "authentication", label = "Authentication"),
+      @ElementTemplate.PropertyGroup(id = "operation", label = "Operation"),
+      @ElementTemplate.PropertyGroup(id = "additionalProperties", label = "Additional properties")
+    },
+    metadata =
+        @ElementTemplate.Metadata(
+            keywords = {
+              "download file from google cloud storage",
+              "upload file to google cloud storage"
```
Comment (chillleader) [inline connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/ObjectStorageConnectorFunction.java:44]: I'd also add `gcs`

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        contentTypeOpt.ifPresent(requestBuilder::contentType);
+
+        return this.createDocument.andThen(DocumentContent::new).apply(requestBuilder.build());
+      } else {
+        return new DownloadResponse.StringContent(outputStream.toString(StandardCharsets.UTF_8));
+      }
+
+    } catch (Exception e) {
+      log.error(
+          "GCS: Failed to download file {} to bucket {}: {}",
```
Comment (chillleader) [inline connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:118]: ~~to~~ **from** bucket

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+              .getService();
+
+      Blob blob = storage.createFrom(blobInfo, uploadObject.document().asInputStream());
+      return new UploadResponse(blob.getBucket(), blob.getName());
+    } catch (IOException ioException) {
+      log.error(
+          "GCS: Failed to upload file {} to bucket {}: {}",
+          fileName,
+          uploadObject.bucket(),
+          ioException.getMessage());
```
Comment (chillleader) [inline connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:88]: nit: consider including the actual exception as the last param to make sure stack trace is also logged

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  }
+
+  private DownloadResponse download(DownloadObject downloadObject) {
+    StorageOptions storageOptions =
+        StorageOptions.newBuilder().setProjectId(downloadObject.project()).build();
+    try (Storage storage = storageOptions.getService()) {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      BlobId blobId = BlobId.of(downloadObject.bucket(), downloadObject.fileName());
+      storage.downloadTo(blobId, outputStream);
```
Comment (chillleader) [inline connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:99]: This puts the file into memory. I recommend checking out the the [`Storage::reader`](https://cloud.google.com/java/docs/reference/google-cloud-storage/latest/com.google.cloud.storage.Storage#com_google_cloud_storage_Storage_reader_com_google_cloud_storage_BlobId_com_google_cloud_storage_Storage_BlobSourceOption____) method to see if it's possible to achieve something like this:

```
try (ReadChannel reader = storage.reader(blobId)) {
    InputStream inputStream = Channels.newInputStream(reader);
    // now pass the input stream directly to the `DocumentCreationRequest`
}
```

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/request/DownloadObject.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            label = "Return document as reference",
+            id = "downloadOperationAsFile",
+            group = "operation",
+            tooltip =
+                "By default, only a reference to the document is returned. If this option is unchecked, the full content of the document is extracted and included in the response.",
+            type = TemplateProperty.PropertyType.Boolean,
+            defaultValueType = TemplateProperty.DefaultValueType.Boolean,
+            defaultValue = "true",
+            binding = @TemplateProperty.PropertyBinding(name = "operation.asFile"))
+        boolean asFile)
```
Comment (chillleader) [inline connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/request/DownloadObject.java:53]: Really great to have this flexibility 👍 

Can we adjust the description a bit to make it clear that:
- If checked, a **Camunda document** is created and its reference is returned
- If not checked, no document is created and the content is passed as is

Also maybe we could rename the field to `asDocument`?

**Review by chillleader — COMMENTED** [review]
Great work 🚀

**Comment by YanaSegal** [comment]
A few more comments:

1. Please remove the hyphen from the template name. 
2. Please don't capitalize in the middle of a sentence unless it is a name. 
Please use this text:
`JSON key of the service account`
`The JSON key of the service account must not be empty.`
`This is the key of the service account in JSON format. See documentation for details.`
image

Thanks!

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+          uploadObject.bucket(),
+          ioException);
+      throw new ConnectorException("Failed to upload file", ioException);
+    }
+  }
+
+  private DownloadResponse download(DownloadObject downloadObject) {
+    StorageOptions storageOptions =
+        StorageOptions.newBuilder().setProjectId(downloadObject.project()).build();
+    try (Storage storage = storageOptions.getService()) {
```
Comment (johnBgood) [inline connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:99]: we don't need credentials?

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            : uploadObject.document().metadata().getFileName();
+    BlobId blobId = BlobId.of(uploadObject.bucket(), fileName);
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+
+    try {
+      ServiceAccountCredentials sac =
+          ServiceAccountCredentials.fromStream(
+              new ByteArrayInputStream(
+                  authentication.getJsonKey().getBytes(StandardCharsets.UTF_8)));
+      Storage storage =
```
[the remark this answers is not in the document]
Comment (ztefanie) [reply connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:77]: According to the google docs example [here](https://cloud.google.com/storage/docs/uploading-objects#storage-upload-object-code-sample) it is not needed. I am a bit confused, because it is done for the download in their example, but Storage does not implement `Closable`. I think i would stick with this though as it follows the documentation, but feel free to disagree and i can change it.

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java (hunk 6 above)
Comment (ztefanie) [reply connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:99]: Good point. Not sure why this worked before :thinking: I added the credentials.

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Test required in (you can check both):
  - [ ] SaaS
  - [ ] SM
- [ ] Create a new test for this feature in our [Pre Release Tests](https://modeler.ultrawombat.com/projects/017779a8-9459-4feb-8b47-eeef8b2717ef--connectors-pre-release-test) for @Szik to be able to test

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

File: connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ztefanie) [reply connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java:74]: ok you are right, i check for this. Not sure how i could miss it :facepalm: I will fix it for upload and download.
